File: python_examples/trunk/src/dump_cspad2x2.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id: dump_cspad2x2.py 2862 2012-02-07 00:54:34Z salnikov@SLAC.STANFORD.EDU $
#
# Description:
#  Pyana user analysis module dump_cspad2x2...
#
#------------------------------------------------------------------------

"""User analysis module for pyana framework.

This software was developed for the LCLS project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@see RelatedModule

@version $Id: dump_cspad2x2.py 2862 2012-02-07 00:54:34Z salnikov@SLAC.STANFORD.EDU $

@author Andy Salnikov
"""

#------------------------------
#  Module's version from SVN --
#------------------------------
__version__ = "$Revision: 2862 $"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#-----------------------------
# Imports for other modules --
#-----------------------------
from pypdsdata import xtc

#----------------------------------
# Local non-exported definitions --
#----------------------------------

# local definitions usually start with _

#---------------------
#  Class definition --
#---------------------
class dump_cspad2x2 (object) :
    """Class whose instance will be used as a user analysis module. """

    #--------------------
    #  Class variables --
    #--------------------
    
    # usual convention is to prefix static variables with s_
    s_staticVariable = 0

    #----------------
    #  Constructor --
    #----------------
    def __init__ ( self, source="" ) :
        """Class constructor takes the name of the data source.

        @param source   data source
        """
        
        self.m_src = source

    #-------------------
    #  Public methods --
    #-------------------

    def beginjob( self, evt, env ) :
 
        config = env.getConfig(xtc.TypeId.Type.Id_Cspad2x2Config, self.m_src)
        if not config:
            return
        
        print "dump_cspad2x2: %s: %s" % (config.__class__.__name__, self.m_src)
        print "  tdi =", config.tdi()
        prot = config.protectionThreshold()
        print "  protectionThreshold: adcThreshold =", prot.adcThreshold, "pixelCountThreshold =", prot.pixelCountThreshold 
        print "  protectionEnable =", config.protectionEnable()
        print "  inactiveRunMode =", config.inactiveRunMode()
        print "  activeRunMode =", config.activeRunMode()
        print "  payloadSize =", config.payloadSize()
        print "  badAsicMask =", config.badAsicMask()
        print "  asicMask =", config.asicMask()
        print "  roiMask =", config.roiMask()
        print "  numAsicsRead =", config.numAsicsRead()
        print "  numAsicsStored =", config.numAsicsStored()
        print "  concentratorVersion =", config.concentratorVersion()
        quad = config.quad()
        print "  quad:"
        print "    shiftSelect =", quad.shiftSelect()
        print "    edgeSelect =", quad.edgeSelect()
        print "    readClkSet =", quad.readClkSet()
        print "    readClkHold =", quad.readClkHold()
        print "    dataMode =", quad.dataMode()
        print "    prstSel =", quad.prstSel()
        print "    acqDelay =", quad.acqDelay()
        print "    intTime =", quad.intTime()
        print "    digDelay =", quad.digDelay()
        print "    ampIdle =", quad.ampIdle()
        print "    injTotal =", quad.injTotal()
        print "    rowColShiftPer =", quad.rowColShiftPer()
        print "    ampReset =", quad.ampReset()
        print "    digCount =", quad.digCount()
        print "    digPeriod =", quad.digPeriod()
        print "    PeltierEnable =", quad.PeltierEnable()
        print "    kpConstant =", quad.kpConstant()
        print "    kiConstant =", quad.kiConstant()
        print "    kdConstant =", quad.kdConstant()
        print "    humidThold =", quad.humidThold()
        print "    setPoint =", quad.setPoint()
        print "    ro =", quad.ro()
        print "    dp =", quad.dp()
        print "    gm =", quad.gm()


    def event( self, evt, env ) :

        elem = evt.get(xtc.TypeId.Type.Id_Cspad2x2Element, self.m_src)
        if not elem :
            print '*** cspad MiniElement information is missing ***'
            return

        # dump information about quadrants
        print "dump_cspad2x2: %s: %s" % (elem.__class__.__name__, self.m_src)
        print "    quadrant: %d" % elem.quad()
        print "    virtual_channel: %s" % elem.virtual_channel()
        print "    lane: %s" % elem.lane()
        print "    tid: %s" % elem.tid()
        print "    acq_count: %s" % elem.acq_count()
        print "    op_code: %s" % elem.op_code()
        print "    seq_count: %s" % elem.seq_count()
        print "    ticks: %s" % elem.ticks()
        print "    fiducials: %s" % elem.fiducials()
        print "    frame_type: %s" % elem.frame_type()
        print "    sb_temp: %s" % map(elem.sb_temp, range(4))

        # image data as 3-dimentional array
        data = elem.data()
        print "    Data shape: %s" % str(data.shape)
        print "    Data:", data

    def endjob( self, env ) :

        logging.info( "dump_cspad2x2.endjob() called" )

File: psana_examples/tags/V00-05-01/src/DumpCsPad2x2.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class DumpCsPad2x2...
//
// Author List:
//      Andrei Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "psana_examples/DumpCsPad2x2.h"

//-----------------
// C/C++ Headers --
//-----------------
#include <algorithm>
#include <iomanip>
#include <iterator>
#include <numeric>
#include <functional>

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "MsgLogger/MsgLogger.h"
#include "psddl_psana/cspad2x2.ddl.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

using namespace psana_examples;
PSANA_MODULE_FACTORY(DumpCsPad2x2)

//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace psana_examples {

//----------------
// Constructors --
//----------------
DumpCsPad2x2::DumpCsPad2x2 (const std::string& name)
  : Module(name)
{
  m_key = configStr("inputKey", "");
  m_src = configSrc("source", "DetInfo(:Cspad2x2)");
}

//--------------
// Destructor --
//--------------
DumpCsPad2x2::~DumpCsPad2x2 ()
{
}

// Method which is called at the beginning of the calibration cycle
void 
DumpCsPad2x2::beginCalibCycle(Event& evt, Env& env)
{
  MsgLog(name(), trace, "in beginCalibCycle()");

  shared_ptr<Psana::CsPad2x2::ConfigV1> config1 = env.configStore().get(m_src);
  if (config1) {
    
    WithMsgLog(name(), info, str) {
      str << "CsPad2x2::ConfigV1:";
      str << "\n  concentratorVersion = " << config1->concentratorVersion();
      str << "\n  protectionEnable = " << config1->protectionEnable();
      str << "\n  protectionThreshold:";
      str << "\n    adcThreshold= " << config1->protectionThreshold().adcThreshold()
          << "\n    pixelCountThreshold= " << config1->protectionThreshold().pixelCountThreshold();
      str << "\n  inactiveRunMode = " << config1->inactiveRunMode();
      str << "\n  activeRunMode = " << config1->activeRunMode();
      str << "\n  tdi = " << config1->tdi();
      str << "\n  payloadSize = " << config1->payloadSize();
      str << "\n  badAsicMask1 = " << config1->badAsicMask();
      str << "\n  asicMask = " << config1->asicMask();
      str << "\n  numAsicsRead = " << config1->numAsicsRead();
      str << "\n  roiMask = " << config1->roiMask();
      str << "\n  numAsicsStored = " << config1->numAsicsStored();
      const Psana::CsPad2x2::ConfigV1QuadReg& quad = config1->quad();
      str << "\n  quad:";
      str << "\n    shiftSelect = " << quad.shiftSelect();
      str << "\n    edgeSelect = " << quad.edgeSelect();
      str << "\n    readClkSet = " << quad.readClkSet();
      str << "\n    readClkHold = " << quad.readClkHold();
      str << "\n    dataMode = " << quad.dataMode();
      str << "\n    prstSel = " << quad.prstSel();
      str << "\n    acqDelay = " << quad.acqDelay();
      str << "\n    intTime = " << quad.intTime();
      str << "\n    digDelay = " << quad.digDelay();
      str << "\n    ampIdle = " << quad.ampIdle();
      str << "\n    injTotal = " << quad.injTotal();
      str << "\n    rowColShiftPer = " << quad.rowColShiftPer();
      str << "\n    ampReset = " << quad.ampReset();
      str << "\n    digCount = " << quad.digCount();
      str << "\n    digPeriod = " << quad.digPeriod();
      str << "\n    PeltierEnable = " << quad.PeltierEnable();
      str << "\n    kpConstant = " << quad.kpConstant();
      str << "\n    kiConstant = " << quad.kiConstant();
      str << "\n    kdConstant = " << quad.kdConstant();
      str << "\n    humidThold = " << quad.humidThold();
      str << "\n    setPoint = " << quad.setPoint();
      str << "\n    digitalPots = " << quad.dp().pots();
      str << "\n    readOnly = shiftTest: " << quad.ro().shiftTest() << " verstion: " << quad.ro().version();
      str << "\n    gainMap = " << quad.gm().gainMap();
    }
    
  }

  shared_ptr<Psana::CsPad2x2::ConfigV2> config2 = env.configStore().get(m_src);
  if (config2) {

    WithMsgLog(name(), info, str) {
      str << "CsPad2x2::ConfigV2:";
      str << "\n  concentratorVersion = " << config2->concentratorVersion();
      str << "\n  protectionEnable = " << config2->protectionEnable();
      str << "\n  protectionThreshold:";
      str << "\n    adcThreshold= " << config2->protectionThreshold().adcThreshold()
          << "\n    pixelCountThreshold= " << config2->protectionThreshold().pixelCountThreshold();
      str << "\n  inactiveRunMode = " << config2->inactiveRunMode();
      str << "\n  activeRunMode = " << config2->activeRunMode();
      str << "\n  runTriggerDelay = " << config2->runTriggerDelay();
      str << "\n  tdi = " << config2->tdi();
      str << "\n  payloadSize = " << config2->payloadSize();
      str << "\n  badAsicMask1 = " << config2->badAsicMask();
      str << "\n  asicMask = " << config2->asicMask();
      str << "\n  numAsicsRead = " << config2->numAsicsRead();
      str << "\n  roiMask = " << config2->roiMask();
      str << "\n  numAsicsStored = " << config2->numAsicsStored();
      const Psana::CsPad2x2::ConfigV2QuadReg& quad = config2->quad();
      str << "\n  quad:";
      str << "\n    shiftSelect = " << quad.shiftSelect();
      str << "\n    edgeSelect = " << quad.edgeSelect();
      str << "\n    readClkSet = " << quad.readClkSet();
      str << "\n    readClkHold = " << quad.readClkHold();
      str << "\n    dataMode = " << quad.dataMode();
      str << "\n    prstSel = " << quad.prstSel();
      str << "\n    acqDelay = " << quad.acqDelay();
      str << "\n    intTime = " << quad.intTime();
      str << "\n    digDelay = " << quad.digDelay();
      str << "\n    ampIdle = " << quad.ampIdle();
      str << "\n    injTotal = " << quad.injTotal();
      str << "\n    rowColShiftPer = " << quad.rowColShiftPer();
      str << "\n    ampReset = " << quad.ampReset();
      str << "\n    digCount = " << quad.digCount();
      str << "\n    digPeriod = " << quad.digPeriod();
      str << "\n    biasTuning = " << quad.biasTuning();
      str << "\n    pdpmndnmBalance = " << quad.pdpmndnmBalance();
      str << "\n    PeltierEnable = " << quad.PeltierEnable();
      str << "\n    kpConstant = " << quad.kpConstant();
      str << "\n    kiConstant = " << quad.kiConstant();
      str << "\n    kdConstant = " << quad.kdConstant();
      str << "\n    humidThold = " << quad.humidThold();
      str << "\n    setPoint = " << quad.setPoint();
      str << "\n    digitalPots = " << quad.dp().pots();
      str << "\n    readOnly = shiftTest: " << quad.ro().shiftTest() << " verstion: " << quad.ro().version();
      str << "\n    gainMap = " << quad.gm().gainMap();
    }

  }
}

// Method which is called with event data
void 
DumpCsPad2x2::event(Event& evt, Env& env)
{
  shared_ptr<Psana::CsPad2x2::ElementV1> elem1 = evt.get(m_src, m_key);
  if (elem1) {

    WithMsgLog(name(), info, str) {
      str << "CsPad2x2::ElementV1:";
      str << "\n  virtual_channel = " << elem1->virtual_channel() ;
      str << "\n  lane = " << elem1->lane() ;
      str << "\n  tid = " << elem1->tid() ;
      str << "\n  acq_count = " << elem1->acq_count() ;
      str << "\n  op_code = " << elem1->op_code() ;
      str << "\n  quad = " << elem1->quad() ;
      str << "\n  seq_count = " << elem1->seq_count() ;
      str << "\n  ticks = " << elem1->ticks() ;
      str << "\n  fiducials = " << elem1->fiducials() ;
      str << "\n  frame_type = " << elem1->frame_type() ;

      str << "\n    sb_temp = " << elem1->sb_temp();

      const ndarray<const int16_t, 3>& data = elem1->data();
      str << "\n    common_mode = [ ";
      for (unsigned i = 0; i != data.shape()[2]; ++ i) {
          str << elem1->common_mode(i) << ' ';
      }
      str << "]";
      str << "\n    data = " << data;
    }
  }

}

} // namespace psana_examples
